api/api.go

```diff
@@ -1,187 +1,136 @@
-		a.HTTPError(w, errors.UnmarshalRequest.Combine(err))
-		return
+// NewAPI returns an *API
+func NewAPI(volplugin Volplugin, hostname string, client *config.Client, global **config.Global) *API {
+	return &API{
+		Volplugin:       volplugin,
+		Hostname:        hostname,
+		Client:          client,
+		Global:          global,
+		Lock:            lock.NewDriver(client),
+		MountCollection: mount.NewCollection(),
+		MountCounter:    mount.NewCounter(),
+		lockStopChans:   map[string]chan struct{}{},
 	}
+}
 
-	parts := strings.SplitN(req.Name, "/", 2)
-	if len(parts) != 2 {
-		a.HTTPError(w, errors.UnmarshalRequest.Combine(errors.InvalidVolume))
-		return
+// RESTHTTPError returns a 500 status with the error.
+func RESTHTTPError(w http.ResponseWriter, err error) {
+	if err == nil {
+		err = errors.Unknown
 	}
 
-	policy, volume := parts[0], parts[1]
+	log.Errorf("Returning HTTP error handling plugin negotiation: %s", err.Error())
+	http.Error(w, err.Error(), http.StatusInternalServerError)
+}
 
-	if policy == "" {
-		a.HTTPError(w, errors.GetPolicy.Combine(errored.Errorf("policy was blank for volume %q", req.Name)))
-		return
-	}
+// Action is a catchall for additional driver functions.
+func Action(w http.ResponseWriter, r *http.Request) {
+	defer r.Body.Close()
+	log.Debugf("Unknown driver action at %q", r.URL.Path)
+	content, _ := ioutil.ReadAll(r.Body)
+	log.Debug("Body content:", string(content))
+	w.WriteHeader(503)
+}
 
-	if volume == "" {
-		a.HTTPError(w, errors.GetVolume.Combine(errored.Errorf("volume was blank for volume %q", req.Name)))
-		return
-	}
+// LogHandler injects a request logging handler if debugging is active. In either event it will dispatch.
+func LogHandler(name string, debug bool, actionFunc func(http.ResponseWriter, *http.Request)) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if debug {
+			buf := new(bytes.Buffer)
+			io.Copy(buf, r.Body)
+			log.Debugf("Dispatching %s with %v", name, strings.TrimSpace(string(buf.Bytes())))
+			var writer *io.PipeWriter
+			r.Body, writer = io.Pipe()
+			go func() {
+				io.Copy(writer, buf)
+				writer.Close()
+			}()
+		}
 
-	if vol, err := a.Client.GetVolume(policy, volume); err == nil && vol != nil {
-		a.HTTPError(w, errors.Exists)
-		return
+		actionFunc(w, r)
 	}
+}
 
-	log.Infof("Creating volume %q", req.Name)
-
-	hostname, err := os.Hostname()
+// GetStorageParameters accepts a Volume API request and turns it into several internal structs.
+func (a *API) GetStorageParameters(uc *Volume) (storage.MountDriver, *config.Volume, storage.DriverOptions, error) {
+	driverOpts := storage.DriverOptions{}
+	volConfig, err := a.Client.GetVolume(uc.Policy, uc.Name)
 	if err != nil {
-		a.HTTPError(w, errors.GetHostname.Combine(err))
-		return
+		return nil, nil, driverOpts, err
 	}
 
-	policyObj, err := a.Client.GetPolicy(policy)
+	driver, err := backend.NewMountDriver(volConfig.Backends.Mount, (*a.Global).MountPath)
```

Comment:
You might want to use `NewDriver()` here. In that way, we will not expose `NewMountDriver`, `NewSanpshotDriver`, etc.